### PR TITLE
feat: Render Perforations inside esv-intersection.

### PR DESCRIPTION
**Pull request by ovineq**
Renders perforations in the esv-intersection based on the ruleset provided here by @ooystein: https://app.zenhub.com/workspaces/wellx-5f89a0db386bba0014989b28/issues/gh/equinor/wellx-designer/222

Related issue: https://github.com/equinor/esv-intersection/issues/484

https://app.zenhub.com/workspaces/wellx-5f89a0db386bba0014989b28/issues/gh/equinor/esv-intersection/484

image

image

image

image

### CODE REVIEW

File: src/datautils/schematicShapeGenerator.ts [hunk 1]
```diff
@@ -601,182 +628,168 @@ export const createComplexRopeSegmentsForPerforation = (
...
-  widestPerfShapeDiameter: number,
+  _perfShape: ComplexRopeSegment,
+  perforationOptions: PerforationOptions,
+) => {
+  ctx.fillStyle = perforationOptions.yellow;
+  ctx.strokeStyle = perforationOptions.yellow;
+
+  const { packingOpacity } = perforationOptions;
+
+  ctx.fillStyle = perforationOptions.yellow;
```
Comment (ovineq) [inline src/datautils/schematicShapeGenerator.ts]: Can be cleaned up

File: src/datautils/schematicShapeGenerator.ts [hunk 2]
```diff
@@ -788,35 +801,18 @@ const createSubkindPerforationTexture = (
...
-  const xy: [number, number] = [0, canvas.height / 2 - diameter / 2];
-  const wh: [number, number] = [canvas.width, diameter];
-  canvasCtx.save();
-  canvasCtx.globalAlpha = packingOpacity;
-  canvasCtx.fillRect(...xy, ...wh);
-  canvasCtx.restore();
+  return { canvas, ctx };
+};
 
+const createTexture = (canvas: HTMLCanvasElement) => {
```
Comment (ovineq) [inline src/datautils/schematicShapeGenerator.ts:813]: Seems to be related to perforations, let's change the naming to be more related.

File: src/datautils/schematicShapeGenerator.ts [hunk 3]
```diff
@@ -788,35 +801,18 @@ const createSubkindPerforationTexture = (
   return texture;
 };
 
-/**
- * Yellow gravel
- * @param perforationOptions
- * @returns
- */
-const createSubkindOpenHoleGravelPackTexture = (perforationOptions: PerforationOptions, widestPerfShapeDiameter: number) => {
+const createCanvas = (perfShape: ComplexRopeSegment, options: PerforationOptions): { canvas: HTMLCanvasElement; ctx: CanvasRenderingContext2D } => {
```
Comment (ovineq) [inline src/datautils/schematicShapeGenerator.ts:802]: Seems to be related to perforations, let's change the naming to be more related.

File: src/datautils/schematicShapeGenerator.ts [hunk 4]
```diff
@@ -828,146 +824,206 @@ const createSubkindOpenHoleGravelPackTexture = (perforationOptions: PerforationO
...
+    otherPerforations: Perforation[],
+    perforationOptions: PerforationOptions,
+  ): Texture => {
+    const { canvas, ctx } = createCanvas(perfShape, perforationOptions);
+
+    const intersectionsWithCasedHoleGravel: boolean = otherPerforations.some(
+      (perf) => isSubkindCasedHoleGravelPack(perf) && intersect(perforation, perf),
+    );
 
-  const { /*fracLineLength,*/ packingOpacity } = perforationOptions;
```
Comment (ovineq) [inline src/datautils/schematicShapeGenerator.ts]: Remove comment

File: src/datautils/schematicShapeGenerator.ts [hunk 5]
```diff
@@ -828,146 +824,206 @@ const createSubkindOpenHoleGravelPackTexture = (perforationOptions: PerforationO
...
+  packing: (perfShape: ComplexRopeSegment, perforationOptions: PerforationOptions): Texture => {
+    const { canvas, ctx } = createCanvas(perfShape, perforationOptions);
+    drawPacking(canvas, ctx, perfShape, perforationOptions);
+    return createTexture(canvas);
+  },
+  fracLines: (perfShape: ComplexRopeSegment, perforationOptions: PerforationOptions) => {
+    const { canvas } = createCanvas(perfShape, perforationOptions);
+    // const noIntersectionsWithSubkindPerforation: boolean = otherPerforations.some(
+    //   (perf) => isSubKindPerforation(perf) && !intersect(perforation, perf),
+    // );
```
Comment (ovineq) [inline src/datautils/schematicShapeGenerator.ts:920]: Remove commented code

File: src/datautils/schematicShapeGenerator.ts [hunk 6]
```diff
@@ -828,146 +824,206 @@ const createSubkindOpenHoleGravelPackTexture = (perforationOptions: PerforationO
...
-  const texture = new Texture(
-    Texture.from(canvas, { wrapMode: WRAP_MODES.CLAMP }).baseTexture,
-    null,
-    new Rectangle(0, 0, canvas.width, canvas.height),
-    null,
-    groupD8.MIRROR_HORIZONTAL,
+export const createPerforationPackingTexture = (
+  perforation: Perforation,
+  perfShape: ComplexRopeSegment,
+  _otherPerforations: Perforation[],
```
Comment (ovineq) [inline src/datautils/schematicShapeGenerator.ts:976]: Unused

File: src/datautils/schematicShapeGenerator.ts [hunk 7]
```diff
@@ -828,146 +824,206 @@ const createSubkindOpenHoleGravelPackTexture = (perforationOptions: PerforationO
...
- * Yellow gravel and fracturation lines.
- * @param perforationOptions
- * @returns
- */
-const createSubkindCasedHoleFracPackTexture = (
+export const createPerforationFracLineTexture = (
   perforation: Perforation,
-  otherPerforations: Perforation[],
-  widestPerfShapeDiameter: number,
+  _otherPerforations: Perforation[],
```
Comment (ovineq) [inline src/datautils/schematicShapeGenerator.ts:994]: Unused

File: src/layers/SchematicLayer.ts [hunk 8]
```diff
@@ -417,19 +422,91 @@ export class SchematicLayer<T extends SchematicData> extends PixiLayer<T> {
...
+        Object.values(perfShapesByDiameter).forEach((perfShapesWithSameDiameter) => {
+          const texture = this.createPerforationPackingTexture(perforation, perfShapesWithSameDiameter[0], perforations, perforationOptions);
+          const rope = this.drawComplexRope(perfShapesWithSameDiameter, texture);
+          this.perforationRopeAndTextureReferences.push({ rope, texture });
+        });
+      });
+
+      fracLines.forEach((perforation) => {
+        const perfShapes = this.createPerforationShape(perforation, casings, holeSizes);
+        // const otherPerforations = perforations.filter((p) => p.id !== perforation.id);
```
Comment (ovineq) [inline src/layers/SchematicLayer.ts:466]: Unused

File: src/layers/SchematicLayer.ts [hunk 9]
```diff
@@ -417,19 +422,91 @@ export class SchematicLayer<T extends SchematicData> extends PixiLayer<T> {
...
+          perfShapesWithSameDiameter.forEach((perfShape) => {
+            const texture = this.createPerforationFracLineTexture(perforation, perfShape, perforations, perforationOptions);
+            const rope = this.drawComplexRope([perfShape], texture);
+            this.perforationRopeAndTextureReferences.push({ rope, texture });
+          });
+        });
+      });
+      spikes.forEach((perforation) => {
+        const perfShapes = this.createPerforationShape(perforation, casings, holeSizes);
+        // const otherPerforations = perforations.filter((p) => p.id !== perforation.id);
```
Comment (ovineq) [inline src/layers/SchematicLayer.ts:488]: Unused

File: src/layers/SchematicLayer.ts [hunk 10]
```diff
@@ -417,19 +422,91 @@ export class SchematicLayer<T extends SchematicData> extends PixiLayer<T> {
...
+      rope.destroy({
+        children: true,
+        texture: true,
+        baseTexture: true,
+      });
+      texture.destroy(true);
+    });
+    this.perforationRopeAndTextureReferences = [];
+
+    // PERFORATION RENDER LOGIC
```
Comment (ovineq) [inline src/layers/SchematicLayer.ts:438]: Remove comment

**Review by ovineq — COMMENTED** [review]
Some things to clean up. Might need to pair-review the rest.

**Comment by ooystein** [comment]
Please update PR name and add description with anything the reviewer needs to know @schoening @ovineq

**Review by ooystein — CHANGES_REQUESTED** [review]
Exited to see this! Just a few simple comments and this should be ready!

**Review by ooystein — APPROVED** [review]
:shipit:

**Pull request merged**